Documentation/proposals/202405-agent-daemonset.md

```diff
@@ -0,0 +1,133 @@
+
+This deployment mode has been implemented in [Google Cloud Managed Service for Prometheus (GMP)](https://github.com/GoogleCloudPlatform/prometheus-engine/), so we have an implementation example to learn from. Also, since this deployment has been tested and proven with GMP’s userbase, we can count on that it can solve a large enough number of use cases.
+
+
+## 2. Pitfalls of the current solution
+
+The current (StatefulSet) deployment brings along the corresponding pitfalls:
+* Load management & scalability: Since one or several high-availability Prometheus Agents are responsible for scraping metrics of the whole cluster, users would need to calculate/estimate the load and scalability of the whole cluster to decide on replicas and sharding strategies. Estimating cluster-wide load and scalability is a much harder task than estimating node-wide load and scalability. Even though they can use helping tools like Horizontal Pod Autoscaler (HPA), that's still additional complexity.
+* Security: Similarly, cluster-wide security is a much bigger problem than node-wide security.
+
+At the moment, the proposed DaemonSet deployment mode doesn’t mean to be a replacement for the current StatefulSet mode. It’s actually a solution for use cases where StatefulSet may not be the best choice. We’ll keep the StatefulSet mode as long as there’s user need for it.
+
+
+## 3. Audience
+
+Users with use cases where:
+* Scraped load is very large or hard to estimate.
+* Scalability is hard to predict.
+* Security is a big concern.
+
+
+## 4. Goals
+
+Provide an MVP version of the DaemonSet deployment of Prometheus Agent to the Audience.
+In specific, the MVP will need to:
+* Allow users to deploy one Prometheus Agent pod per node.
+* Allow users to restrict which set of nodes they want to deploy Prometheus Agent, if desired.
+* Allow each Prometheus Agent pod to scrape from the pods on the same node (PodMonitor support).
+
+
+## 5. Non-Goals
+
+The non-goals are the features that are not easy to implement and require more investigation. We will need to investigate whether there are actual user needs for them, if yes, then how to best implement them. We’ll handle these after the MVP.
+* ServiceMonitor support: There's a performance issue regarding this feature. Since each Prometheus Agent running on a node requires one watch, making all Prometheus Agent pods watch all endpoints will put a huge stress on Kubernetes API server. This is the main reason why GMP hasn’t supported this, even though there are user needs stated in some issues ([#362](https://github.com/GoogleCloudPlatform/prometheus-engine/issues/362), [#192](https://github.com/GoogleCloudPlatform/prometheus-engine/issues/192)). However, as discussed with Danny from GMP [here](https://github.com/GoogleCloudPlatform/prometheus-engine/issues/192#issuecomment-2028850846), ServiceMonitor support based on EndpointSlice seems like a viable approach. We’ll investigate this further after the MVP.
```

Comment:
This is a non-goal but it'd be nice to attack this if we ever finish our planned goals before the program ends.